File: src/main/java/org/jukeboxmc/network/Network.java
package org.jukeboxmc.network;

import com.nukkitx.protocol.bedrock.*;
import com.nukkitx.protocol.bedrock.v575.Bedrock_v575;
import org.jetbrains.annotations.NotNull;
import org.jukeboxmc.Server;
import org.jukeboxmc.player.PlayerConnection;

import java.net.InetSocketAddress;
import java.util.HashSet;
import java.util.Set;
import java.util.function.Consumer;
import java.util.function.Predicate;

/**
 * @author LucGamesYT
 * @version 1.0
 */
public class Network implements BedrockServerEventHandler {

    public static final BedrockPacketCodec CODEC = Bedrock_v575.V575_CODEC;

    private final Server server;
    private final InetSocketAddress inetSocketAddress;
    private final @NotNull BedrockPong bedrockPong;
    private final @NotNull BedrockServer bedrockServer;

    private final Set<PlayerConnection> connections = new HashSet<>();

    private final @NotNull Predicate<PlayerConnection> removePredicate;
    private final @NotNull Consumer<PlayerConnection> updater;

    public Network( Server server, InetSocketAddress inetSocketAddress ) {
        this.server = server;
        this.inetSocketAddress = inetSocketAddress;
        this.bedrockPong = new BedrockPong();
        this.bedrockServer = new BedrockServer( inetSocketAddress );
        this.bedrockServer.setHandler( this );

        this.removePredicate = PlayerConnection::isClosed;
        this.updater = PlayerConnection::update;

        try {
            this.bedrockServer.bind().join();
            this.server.getLogger().info( "Server started successfully at " + this.inetSocketAddress.getHostString() + ":" + this.inetSocketAddress.getPort() + "!" );
        } catch ( Exception e ) {
            this.server.getLogger().error( "Could not start server! Is there already running something on this port?", e );
        }
    }

    @Override
    public BedrockPong onQuery( @NotNull InetSocketAddress inetSocketAddress ) {
        this.bedrockPong.setEdition( "MCPE" );
        this.bedrockPong.setGameType( this.server.getGameMode().getIdentifier() );
        this.bedrockPong.setMotd( this.server.getMotd() );
        this.bedrockPong.setSubMotd( this.server.getSubMotd() );
        this.bedrockPong.setPlayerCount( this.server.getOnlinePlayers().size() );
        this.bedrockPong.setMaximumPlayerCount( this.server.getMaxPlayers() );
        this.bedrockPong.setIpv4Port( this.inetSocketAddress.getPort() );
        this.bedrockPong.setNintendoLimited( false );
        this.bedrockPong.setProtocolVersion( CODEC.getProtocolVersion() );
        this.bedrockPong.setVersion( CODEC.getMinecraftVersion() );
        return this.bedrockPong;
    }

    @Override
    public boolean onConnectionRequest( @NotNull InetSocketAddress address ) {
        return this.server.getFinishedState().get() && this.server.getRunningState().get();
    }

    @Override
    public void onSessionCreation( @NotNull BedrockServerSession bedrockServerSession ) {
        try {
            this.server.addPlayer( this.addPlayer( new PlayerConnection( this.server, bedrockServerSession ) ).getPlayer() );
        } catch ( Throwable e ) {
            e.printStackTrace();
        }
    }

    private synchronized PlayerConnection addPlayer( PlayerConnection playerConnection ) {
        this.connections.add( playerConnection );
        return playerConnection;
    }

    public synchronized void update() {
        this.connections.removeIf( this.removePredicate );
        this.connections.forEach( this.updater );
    }

    public @NotNull BedrockServer getBedrockServer() {
        return this.bedrockServer;
    }

    public Server getServer() {
        return this.server;
    }
}

File: src/main/java/org/jukeboxmc/network/Network.kt
package org.jukeboxmc.network

import io.netty.bootstrap.ServerBootstrap
import io.netty.channel.Channel
import io.netty.channel.nio.NioEventLoopGroup
import io.netty.channel.socket.nio.NioDatagramChannel
import org.cloudburstmc.netty.channel.raknet.RakChannelFactory
import org.cloudburstmc.netty.channel.raknet.config.RakChannelOption
import org.cloudburstmc.protocol.bedrock.BedrockPong
import org.cloudburstmc.protocol.bedrock.BedrockServerSession
import org.cloudburstmc.protocol.bedrock.codec.BedrockCodec
import org.cloudburstmc.protocol.bedrock.codec.v575.Bedrock_v575
import org.cloudburstmc.protocol.bedrock.netty.initializer.BedrockServerInitializer
import org.jukeboxmc.Server
import org.jukeboxmc.player.PlayerConnection
import java.net.InetSocketAddress
import java.util.function.Consumer

/**
 * @author LucGamesYT
 * @version 1.0
 */
class Network(val server: Server, val inetSocketAddress: InetSocketAddress) {
    private val bedrockPong: BedrockPong = BedrockPong().apply {
        edition("MCPE")
        gameType(server.gameMode.identifier)
        motd(server.motd)
        subMotd(server.subMotd)
        playerCount(server.onlinePlayers.size)
        maximumPlayerCount(server.maxPlayers)
        ipv4Port(inetSocketAddress.port)
        nintendoLimited(false)
        protocolVersion(CODEC.protocolVersion)
        version(CODEC.minecraftVersion)
    }
    private val connections: MutableSet<PlayerConnection> = HashSet()
    private val updater: Consumer<PlayerConnection> =
        Consumer<PlayerConnection> { obj: PlayerConnection -> obj.update() }

    private val channels: MutableList<Channel> = mutableListOf()

    init {
        try {
            val bootstrap: ServerBootstrap = ServerBootstrap()
                .channelFactory(RakChannelFactory.server(NioDatagramChannel::class.java)) // TODO: epoll, kqueue
                .group(NioEventLoopGroup())
//                .option(RakChannelOption.RAK_HANDLE_PING, true) // TODO: dynamic
                .option(RakChannelOption.RAK_ADVERTISEMENT, bedrockPong.toByteBuf())
                .childHandler(object : BedrockServerInitializer() {
                    override fun initSession(bedrockServerSession: BedrockServerSession) {
                        val playerConnection = addPlayer(PlayerConnection(server, bedrockServerSession))
                        server.addPlayer(playerConnection.player)
                    }
                })
            channels.add(
                bootstrap.bind(inetSocketAddress)
                    .awaitUninterruptibly()
                    .channel(),
            )
            server.logger.info("Server started successfully at " + this.inetSocketAddress.hostString + ":" + this.inetSocketAddress.port + "!")
        } catch (e: Exception) {
            server.logger.error("Could not start server! Is there already running something on this port?", e)
        }
    }

    fun onQuery(inetSocketAddress: InetSocketAddress): BedrockPong {
//        bedrockPong.edition = "MCPE"
//        bedrockPong.gameType = server.gameMode.identifier
//        bedrockPong.motd = server.motd
//        bedrockPong.subMotd = server.subMotd
//        bedrockPong.playerCount = server.onlinePlayers.size
//        bedrockPong.maximumPlayerCount = server.maxPlayers
//        bedrockPong.ipv4Port = this.inetSocketAddress.port
//        bedrockPong.isNintendoLimited = false
//        bedrockPong.protocolVersion = CODEC.protocolVersion
//        bedrockPong.version = CODEC.minecraftVersion
        return bedrockPong.apply {
            edition("MCPE")
            gameType(server.gameMode.identifier)
            motd(server.motd)
            subMotd(server.subMotd)
            maximumPlayerCount(server.onlinePlayers.size)
            maximumPlayerCount(server.maxPlayers)
            ipv4Port(inetSocketAddress.port)
            nintendoLimited(false)
            protocolVersion(CODEC.protocolVersion)
            version(CODEC.minecraftVersion)
        }
    }

    @Synchronized
    private fun addPlayer(playerConnection: PlayerConnection): PlayerConnection {
        connections.add(playerConnection)
        return playerConnection
    }

    @Synchronized
    fun update() {
        connections.removeIf { obj: PlayerConnection -> obj.isClosed }
        connections.forEach(updater)
    }

    fun shutdown() {
        val playerConnectionIterator = connections.iterator()
        while (playerConnectionIterator.hasNext()) {
            val playerConnection = playerConnectionIterator.next()
            playerConnection.disconnect("Server closed") // TODO: configurable
        }
        channels.forEach(Channel::close)
    }

    companion object {
        val CODEC: BedrockCodec = Bedrock_v575.CODEC
    }
}
